libsolidity/analysis/TypeChecker.cpp

```diff
@@ -113,18 +113,24 @@ void TypeChecker::checkDoubleStorageAssignment(Assignment const& _assignment)
 	size_t storageByteArrayPushes = 0;
 	size_t storageByteAccesses = 0;
 	auto count = [&](TupleExpression const& _lhs, TupleType const& _rhs, auto _recurse) -> void {
-		TupleType const& lhsType = dynamic_cast<TupleType const&>(*type(_lhs));
-		TupleExpression const* lhsResolved = dynamic_cast<TupleExpression const*>(resolveOuterUnaryTuples(&_lhs));
+		auto& lhsType = dynamic_cast<TupleType const&>(*type(_lhs));
+		auto* lhsResolved = dynamic_cast<TupleExpression const*>(resolveOuterUnaryTuples(&_lhs));
+
+		if (!lhsResolved) {
+			m_errorReporter.reportErrors("lhsResolved is null");
+			return;
+		}
 
 		if (lhsType.components().size() != _rhs.components().size() || lhsResolved->components().size() != _rhs.components().size())
 		{
 			solAssert(m_errorReporter.hasErrors(), "");
 			return;
 		}
 
+
 		for (auto&& [index, componentType]: lhsType.components() | ranges::views::enumerate)
 		{
-			if (ReferenceType const* ref = dynamic_cast<ReferenceType const*>(componentType))
+			if (auto* ref = dynamic_cast<ReferenceType const*>(componentType))
```

Comment:
Our convention is to keep the explicit `const` when using `auto` for clarity:

```suggestion
			if (auto const* ref = dynamic_cast<ReferenceType const*>(componentType))
```

Reply:
Also, I think we should avoid doing random code cleanup bugfix. This distracts from the actual fix - most of the changes in the PR seems to be completely unrelated to the bug. Could you limit the PR to just the bugfix and submit any unrelated refactor and comment tweaks as a separate PR?